### PR TITLE
Fix the scheduling decision for multiple currentStates

### LINKED ISSUE #461: Investigate FixedTargetTaskAssignmentCalculator
Right now the targeted assignment for targeted jobs relies on jobContext which might be different from Currentstate.

### LINKED ISSUE #922: Continuous wrong assignment of targeted tasks
There are scenarios where there are multiple currentstates and previous assignment existed in ZK for one partition. This will confused the controller for the task scheduling stage. In this case, since paMap has only place for each task, the final result of scheduling is not deterministic. In this case, tasks can be stuck in INIT -> RUNNING and RUNNING -> DROPPED on the wrong instance.

### PR DESCRIPTION
**Pull request by alirezazamani**
### Issues
- [x] My PR addresses the following Helix issues and references them in the PR title:
Fixes #461
Fixes #922

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
This commit specifically targets the scenario where there are multiple currentstates and previous assignment existed for one partition. In this case, since paMap has only one field for the mapping, the final result of scheduling is not deterministic. In this case, tasks can be stuck in INIT -> RUNNING and RUNNING -> DROPPED on the wrong instance. This PR fixes this scenario.

### Tests
- [x] The following tests are written for this issue:
TestTargetedTaskStateChange
TestTaskSchedulingTwoCurrentStates

- [x] The following is the result of the "mvn test" command on the appropriate module:
Test Result 1: mvn test
Failed tests:
  TestWorkflowTermination.testWorkflowPausedTimeout:170->verifyWorkflowCleanup:257 expected:<true> but was:<false>

Tests run: 1094, Failures: 1, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:13 h
[INFO] Finished at: 2020-03-30T17:11:26-07:00
[INFO] ------------------------------------------------------------------------

The failed test has passed when I ran it individually.

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 1]
```diff
@@ -115,6 +115,12 @@ public void updatePreviousAssignedTasksStatus(
         TaskPartitionState currState = updateJobContextAndGetTaskCurrentState(currStateOutput,
             jobResource, pId, pName, instance, jobCtx, jobTgtState);
 
+        if (!instance.equals(jobCtx.getAssignedParticipant(pId))) {
+          LOG.warn(String.format(
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java]: I suggest using  {}-style parameters.

File: helix-core/src/test/java/org/apache/helix/task/TestTargetedTaskStateChange.java [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+  private ResourceAssignment preparePreviousAssignment(String instance, String state) {
+    ResourceAssignment prevAssignment = new ResourceAssignment(JOB_NAME);
+    Map<String, String> replicaMap = new HashMap<>();
+    replicaMap.put(instance, state);
+    Partition taskPartition = new Partition(JOB_NAME + "_" + PARTITION_NAME);
+    prevAssignment.addReplicaMap(taskPartition, replicaMap);
+    return prevAssignment;
+  }
+
+  private class Mock {
```
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/task/TestTargetedTaskStateChange.java]: Can we have a more descriptive name for this class?

Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/task/TestTargetedTaskStateChange.java]: Done.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 1 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java]: Done.

**Comment by alirezazamani** [comment]
This PR is ready to be merged, approved by @dasahcc 

FInal commit message:
Fix the scheduling decision for multiple currentStates and prevAssignment

In this commit, the problem of scheduling and dropping the tasks
on the slave node has been addressed.
Multiple tests have been added.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestTaskSchedulingTwoCurrentStates.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * Test to check if targeted tasks correctly get assigned and also if cancel messages are not being
+ * sent when there are two CurrentStates.
+ */
+public class TestTaskSchedulingTwoCurrentStates extends TaskTestBase {
+  private final String DATABASE = WorkflowGenerator.DEFAULT_TGT_DB;
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestTaskSchedulingTwoCurrentStates.java:67]: "private static final String" for constants?

File: helix-core/src/test/java/org/apache/helix/integration/task/TestTaskSchedulingTwoCurrentStates.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+    // Stop participants that have been started in super class
+    for (int i = 0; i < _numNodes; i++) {
+      super.stopParticipant(i);
+    }
+
+    // Check that participants are actually stopped
+    for (int i = 0; i < _numNodes; i++) {
+      Assert.assertFalse(_participants[i].isConnected());
+    }
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestTaskSchedulingTwoCurrentStates.java:88]: Possible to merge these two loops?

File: helix-core/src/test/java/org/apache/helix/task/TestTargetedTaskStateChange.java [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+
+  private JobContext prepareJobContext(String instance) {
+    Set<Integer> _taskPartitionSet;
+    Map<Integer, TaskPartitionState> _taskPartitionStateMap;
+    Map<Integer, String> _partitionToTaskIDMap;
+    Map<Integer, String> _taskToInstanceMap;
+    _taskPartitionSet = new HashSet<>();
+    _taskPartitionStateMap = new HashMap<>();
+    _partitionToTaskIDMap = new HashMap<>();
+    _taskToInstanceMap = new HashMap<>();
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/task/TestTargetedTaskStateChange.java:216]: Why do these local variables have underscores?

File: helix-core/src/test/java/org/apache/helix/task/TestTargetedTaskStateChange.java [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+    private CurrentStateOutput _currentStateOutput = prepareCurrentState(masterInstance,
+        slaveInstance, TaskPartitionState.RUNNING.name(), TaskPartitionState.RUNNING.name());
+    private CurrentStateOutput _currentStateOutput2 =
+        prepareCurrentState2(masterInstance, TaskPartitionState.RUNNING.name());
+    private ResourceAssignment _resourceAssignment =
+        preparePreviousAssignment(slaveInstance, TaskPartitionState.RUNNING.name());
+    private ResourceAssignment _resourceAssignment2 =
+        preparePreviousAssignment(slaveInstance, TaskPartitionState.DROPPED.name());
+    private TaskDataCache _taskDataCache = mock(TaskDataCache.class);
+    private RuntimeJobDag _runtimeJobDag = mock(RuntimeJobDag.class);
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/task/TestTargetedTaskStateChange.java:325]: Let's be consistent with underscores here as well? some have them, others don't.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestTaskSchedulingTwoCurrentStates.java (hunk 3 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestTaskSchedulingTwoCurrentStates.java]: Done.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestTaskSchedulingTwoCurrentStates.java (hunk 4 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestTaskSchedulingTwoCurrentStates.java]: Done.

File: helix-core/src/test/java/org/apache/helix/task/TestTargetedTaskStateChange.java (hunk 5 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/task/TestTargetedTaskStateChange.java]: I removed them. Realized we do not need them.

File: helix-core/src/test/java/org/apache/helix/task/TestTargetedTaskStateChange.java (hunk 6 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/task/TestTargetedTaskStateChange.java]: Good point. Fixed.

**Pull request merged**